examples/todos.rs

```diff
@@ -32,8 +32,9 @@ impl Application for Todos {
             }
             Message::CreateTask => {
                 if !self.input_value.is_empty() {
-                    self.tasks.push(Task::new(self.input_value.clone()));
-                    self.input_value = String::new();
+                    // Move the user's input into a task, replacing input_value with an empty string.
+                    let task = Task::new(std::mem::take(&mut self.input_value));
+                    self.tasks.push(task);
```

Comment:
I think the original version is clearer. I'd bet the compiler will optimize the clone away.

However, It'd be nice if Rust could figure this out by itself:

```rust
self.tasks.push(Task::new(self.input_value));
self.input_value = String::new();
```

I don't see why this shouldn't work. Do you know if there is a related RFC for this? Maybe it's just very hard to implement correctly?

Reply:
> I'd bet the compiler will optimize the clone away.

I'm unsure what optimizations like those are permitted around non-copy heap allocated types, but it currently does not optimize away https://rust.godbolt.org/z/AsnV2q

> Maybe it's just very hard to implement correctly?

It would likely complicate move and drop analysis, for both the compiler and users. Currently, when `input_value` moves out, `self` is invalidated, and during assignment, the previous non-copy value is dropped.

That said,

> I think the original version is clearer.

I agree, and now think its more correct. By cloning, `Task`'s string will then have the exact capacity it needs, and if we also use `clear`, `Application` will can re-use `input_value`'s current buffer.

```rust
self.tasks.push(Task::new(self.input_value.clone()));
self.input_value.clear()
```

Thus the use of take/replace was not a good choice for multiple reasons, and clarity in example code is key :)

Reply:
> I'm unsure what optimizations like those are permitted around non-copy heap allocated types, but it currently does not optimize away https://rust.godbolt.org/z/AsnV2q

Oh, wow! I didn't know about Compiler Explorer. This is amazing!

> I agree, and now think its more correct. [...]

Yeah, I like the `clear` version better than the original. It's great to see the simplest solution is also the best one.